serde_list/serde_list_macros/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+// SPDX-FileCopyrightText: 2024 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+#![recursion_limit = "4096"]
+
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use punctuated::Punctuated;
+use quote::quote;
+use syn::*;
+
+#[proc_macro_derive(ExternallyTagged)]
+pub fn derive_externally_tagged(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+    let enum_name = ast.ident;
+
+    let Data::Enum(data) = ast.data else {
+        panic!();
+    };
+
+    let mut valid = false;
+    for attr in &ast.attrs {
+        if attr.path().is_ident("repr") {
+            if attr.parse_args::<Ident>().unwrap() == "u8" {
+                valid = true;
+            }
+        }
+    }
+    if !valid {
+        panic!("ExternallyTagged requires #[repr(u8)]")
+    }
+
+    let mut field_num_impls = Vec::new();
+    let mut field_serialize_impls = Vec::new();
+
+    for variant in data.variants {
+        let variant_name = variant.ident;
+
+        let fields = match variant.fields {
+            Fields::Named(fields_named) => fields_named.named,
+            Fields::Unnamed(_fields_unnamed) => panic!(),
+            Fields::Unit => Punctuated::new(),
+        };
+
+        let mut names = Vec::new();
+        let mut serialized = Vec::new();
+
+        for field in &fields {
+            let Some(field_name) = &field.ident else {
+                panic!();
+            };
+
+            names.push(field_name);
+            serialized.push(quote! { state.serialize_element(#field_name)?; });
+        }
+
+        let num = names.len();
+
+        field_num_impls.push(quote! {
+            Self::#variant_name { .. } => { #num },
+        });
+
+        field_serialize_impls.push(quote! {
+            Self::#variant_name { #(#names),* } => { #(#serialized)* },
+        });
+    }
+
+    quote! {
+        impl ExternallyTagged for #enum_name {
+            // https://doc.rust-lang.org/reference/items/enumerations.html?search=#pointer-casting
+            fn discriminant(&self) -> u8 {
+                // This is safe if the enum has repr(u8)
+                let pointer = self as *const Self as *const u8;
+                unsafe { *pointer }
```

Comment:
We should really try to avoid `unsafe`. `tls_codec_derive` runs into the same problem which is why it doesn't support derivation for the `Custom(u8)` style of enums. Maybe it's okay to require the user to give us some hints. For example, in OpenMLS we implement `From<u8>` and `Into<u8>` traits respectively. Do you think that would be an option?